src/unity/python/turicreate/data_structures/sarray.py

```diff
@@ -4399,6 +4399,92 @@ def cumulative_var(self):
         agg_op = "__builtin__cum_var__"
         return SArray(_proxy = self.__proxy__.builtin_cumulative_aggregate(agg_op))
 
+    def filter_by(self, values, exclude=False):
+        
+        """
+        Filter an SArray by values inside an iterable object. Result is an SArray that
+        only includes (or excludes) the values in the given ``values`` :class:`~turicreate.SArray`.
+        If ``values`` is not an SArray, we attempt to convert it to one before filtering.
+        
+        Parameters
+        ----------
+        values : SArray | list | numpy.ndarray | pandas.Series | str
+        The values to use to filter the SArray. The resulting SArray will
+        only include rows that have one of these values in the given
+        column.
+        
+        exclude : bool
+        If True, the result SArray will contain all rows EXCEPT those that
+        have one of the ``values``.
+        
+        Returns
+        -------
+        out : SArray
+        The filtered SArray.
+        
+        Examples
+        --------
+        >>> sa = SArray(['dog', 'cat', 'cow', 'horse'])
+        >>> sa.filter_by(['cat', 'hamster', 'dog', 'fish', 'bird', 'snake'])
+        dtype: str
+        Rows: 2
+        ['dog', 'cat']
+        
+        >>> sa.filter_by(['cat', 'hamster', 'dog', 'fish', 'bird', 'snake'], exclude=True)
+        dtype: str
+        Rows: 2
+        ['horse', 'cow']
+        """
+    
+        from .sframe import SFrame as _SFrame
+        
+        column_name = 'sarray'
+        
+        #convert values to SArray
+        if type(values) is not SArray:
+            # If we were given a single element, try to put in list and convert
+            # to SArray
+            if not _is_non_string_iterable(values):
+                values = [values]
+            values = SArray(values)
+    
+        #convert values to SFrame
+        value_sf = _SFrame()
+        value_sf.add_column(values, column_name, inplace=True)
+        given_type = value_sf.column_types()[0] #value column type
+        
+        existing_type = self.dtype
+        sarray_sf = _SFrame()
+        sarray_sf.add_column(self, column_name, inplace=True)
+        
+        if given_type != existing_type:
+            raise TypeError("Type of given values does not match type of the SArray column")
+        
+        # Make sure the values list has unique values, or else join will not
+        # filter.
+        value_sf = value_sf.groupby(column_name, {})
+        print(value_sf)
+        
+        with cython_context():
+            if exclude:
+                id_name = "id"
```

Comment:
You don't need to worry about the row number column having a name collision because you just created `value_sf`; you know it doesn't have a column named `id` which is the default column name value for `add_row_number`.

Remove `id_name` variable altogether to make the code simpler.

Reply:
The way I understood left join here is that it returns an SFrame with ids mapped to None for values not in the caller SFrame column.  

tmp: +--------+------+
| sarray |  id  |
+--------+------+
|   2    |  0   |
|   3    |  1   |
|   4    |  4   |
|   7    |  5   |
|   1    | None |
|   5    | None |
|   6    | None |
+--------+------+

Without the row number we don't get any information back to retrieve from:

tmp: +--------+
| sarray |
+--------+
|   1    |
|   2    |
|   3    |
|   4    |
|   5    |
|   6    |
|   7    |
+--------+
[7 rows x 1 columns]

Please correct me if I'm missing something.

Reply:
Right, you still need to call `row_number()`. However you don't need to pass `row_number()` a parameter because you're fine with it using the default value for the new column name that it creates; you know that default value will not be the same as any column name already in the SFrame.

If you want to still pass in a name, that's fine. But please remove the comment about making sure it's unique and remove the `id_name += "1"` line.

Reply:
Thank you for the feedback. I've removed the comment and the id_name += "1" line.